### PR TITLE
[GEP-21] Handle Seed IPFamilies (IPv6)

### PR DESCRIPTION
**Pull request by breuerfelix**
**How to categorize this PR?**
/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
This is part of GEP-21. Implementation of the local IPv6 Seeda

**Which issue(s) this PR fixes**:
Part of #7051 

**Special notes for your reviewer**:
Loki is not yet IPv6 ready until https://github.com/grafana/dskit/pull/185 is merged.

**Release note**:
```feature developer
Gardener's local setup now supports bootstrapping a Seed with IPv6 single-stack networking using `make gardener-up IPFAMILY=ipv6`. See the [documentation](https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally.md) for more detailed steps.
```

### CODE REVIEW

File: charts/images.yaml
Comment (rfranzke) [inline charts/images.yaml]: Can you extend https://github.com/gardener/gardener/blob/676ec251d1381932940be9b31f8e9e781aa62534/charts/images.yaml#L1-L5 with information that Docker Hub should be avoided because it doesn't support IPv6?

File: pkg/features/features.go
Comment (rfranzke) [inline pkg/features/features.go]: I would suggest to drop this commit, please. So far, we preferred individual feature gates per component/binary, and there should be no good reason to change this now. At least not as part of this PR. If you still think it's valuable, please trigger a separate/dedicated discussion about it via opening an according issue.

File: hack/local-development/generate-webhook-ca-bundle
Comment (rfranzke) [inline hack/local-development/generate-webhook-ca-bundle]: The formatting seems broken here

File: hack/local-development/generate-webhook-ca-bundle
Comment (rfranzke) [inline hack/local-development/generate-webhook-ca-bundle]: Add `.sh` suffix to filename

File: example/20-componentconfig-gardenlet.yaml
Comment (rfranzke) [inline example/20-componentconfig-gardenlet.yaml:117]: I don't think this should be enabled in this example file.

```suggestion
  IPv6SingleStack: false
```

File: cmd/gardenlet/app/options.go
Comment (rfranzke) [inline cmd/gardenlet/app/options.go]: Why was this moved to `validate()`? This sounds wrong - nothing is validated here. I would suggest to move it back. Now it's also inconsistent compared to how the feature gates are handled in the other binaries.

File: example/gardener-local/gardenlet/values.yaml
Comment (rfranzke) [inline example/gardener-local/gardenlet/values.yaml:45]: Same here - doesn't this mean IPv6 single stack will be enabled in all local Gardener environments?

File: pkg/operation/botanist/component/istio/charts/istio/istio-ingress/templates/envoy-filter.yaml
Comment (rfranzke) [inline pkg/operation/botanist/component/istio/charts/istio/istio-ingress/templates/envoy-filter.yaml:39]: cc/FYI @ScheererJ 
similar changes:
- https://github.com/gardener/gardener/pull/7561/commits/ad692ab6ec369c43d8e5216a5372610331479d70#diff-54653513c9d30c5e6bd47a82351cdb3b2e67a04f6e29275ca70e3db381c9f582
- https://github.com/gardener/gardener/pull/7561/commits/ad692ab6ec369c43d8e5216a5372610331479d70#diff-c54bf915e534f9bdb05b19b4c9be5e5a3aa8540703eb8426b052d72128032214

File: example/20-componentconfig-gardener-admission-controller-ipv6.yaml
Comment (rfranzke) [inline example/20-componentconfig-gardener-admission-controller-ipv6.yaml]: Why not `"[::]"` here as well?

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (rfranzke) [inline example/gardener-local/gardenlet/values-ipv6.yaml]: Is this needed? I don't see anything IPv6-specific in here?

File: example/provider-local/seed-kind/local-ipv6/seed.yaml
Comment (rfranzke) [inline example/provider-local/seed-kind/local-ipv6/seed.yaml]: Would it be possible to work with patches to prevent copying the entire manifest? Like https://github.com/gardener/gardener/tree/master/example/provider-local/seed-kind2/local?

File: hack/local-development/common/helpers
Comment (rfranzke) [inline hack/local-development/common/helpers]: Sorry?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/features/features.go
Comment (breuerfelix) [reply pkg/features/features.go]: This PR does not work if this commit is not included here unfortunately...
The problem: while validating the config.yaml for the gardenlet, we use the Seed validation since there is a seed object in the config that configures the gardenlet. In this config there is also the definition on which features gates should be activated and which not.
The validation of the Seed object requires the SingleStack feature gate to be activated but we haven't yet activated the gate since we are still validating the config.yaml, thus we haven't applied it.
Would you suggest creating another PR on which this one here depends?

File: example/20-componentconfig-gardenlet.yaml
Comment (breuerfelix) [reply example/20-componentconfig-gardenlet.yaml:117]: It is enabled in the apiserver example per default aswell, thats why i made it here accordingly. Should i disable it in both cases?

File: cmd/gardenlet/app/options.go
Comment (breuerfelix) [reply cmd/gardenlet/app/options.go]: I described the problem in the other comment. TLDR; during the validation of the seed object inside the config, we check if the feature gate is enabled or not.

File: example/gardener-local/gardenlet/values.yaml
Comment (breuerfelix) [reply example/gardener-local/gardenlet/values.yaml:45]: yes, but this just means you can provide IPv6 addresses and Gardener validates them. IPv4 does also work.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (breuerfelix) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: `server: https://kubernetes.default.svc:443` this line changes. Not the internal kubernetes Service gets used for access. cc @einfachnuralex

File: example/20-componentconfig-gardener-admission-controller-ipv6.yaml
Comment (breuerfelix) [reply example/20-componentconfig-gardener-admission-controller-ipv6.yaml]: somehow the webhook gets handled differently and this will throw the error `listen tcp: address [[::]]:2721: missing port in address` because it adds the braces on its own, not like the other servers

File: pkg/operation/botanist/component/istio/charts/istio/istio-ingress/templates/envoy-filter.yaml
Comment (ScheererJ) [reply pkg/operation/botanist/component/istio/charts/istio/istio-ingress/templates/envoy-filter.yaml:39]: `portNumber` should be sufficient indeed. Otherwise we would need to duplicate the section with both an IPv4 and an IPv6 address. This approach seems better.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (ScheererJ) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: Why was it necessary to change from `https://gardener-local-control-plane:6443` to `https://kubernetes.default.svc:443`?

File: example/20-componentconfig-gardenlet.yaml
Comment (rfranzke) [reply example/20-componentconfig-gardenlet.yaml:117]: For the API server, it is only used for validation purposes and only when the `ipFamily=IPv6` was set. Hence, it didn't cause "harm" so far. Perhaps it's indeed reasonable to disable it everywhere?

File: example/20-componentconfig-gardener-admission-controller-ipv6.yaml
Comment (rfranzke) [reply example/20-componentconfig-gardener-admission-controller-ipv6.yaml]: Why is this? 😕 That's a bit odd and probably something we should clean up.

File: example/gardener-local/gardenlet/values.yaml
Comment (rfranzke) [reply example/gardener-local/gardenlet/values.yaml:45]: https://github.com/gardener/gardener/pull/7561#discussion_r1119256147 is related. Maybe yes, maybe the feature gate will lead to behavioural changes in the future. Perhaps it's safer to disable it by default even in the local environments.

File: pkg/features/features.go
Comment (rfranzke) [reply pkg/features/features.go]: Hold your horses for now - as said above, I don't even know whether we want this. I believe it requires a discussion first.
Where is this validation code that you mentioned which requires all this? I tried to find it in the PR/commits, but wasn't successful. Maybe we can find a solution without it or at least a compromise to not block this PR.

Comment (breuerfelix) [reply pkg/features/features.go]: Okay i will open a discussion for this :)
The respective code is here https://github.com/gardener/gardener/blob/931b37c85f6ea6ec5c5e898a3e9c4235b628db7d/pkg/apis/core/validation/utils.go#L192 and it got introduced in an earlier PR where all IPv6 types got introduced (https://github.com/gardener/gardener/pull/7288/files).
The validation code was meant to be used in the apiserver to validate seed objects, but as stated, the gardenlet got a seed object inside its startup configuration.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (breuerfelix) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: We debugged this issue a little bit more yesterday and it basically boils down to this kind issue i opened https://github.com/kubernetes-sigs/kind/issues/3114 
We don't really know why it does not resolve the kind host when ipv6 is enabled as it works in ipv4 perfectly.

File: example/20-componentconfig-gardener-admission-controller-ipv6.yaml
Comment (breuerfelix) [reply example/20-componentconfig-gardener-admission-controller-ipv6.yaml]: We just concatenated host and port together. I updated the PR and use the `net` package which the controller-runtime also uses to put host and port together for the webhook server. Now they all use `::` as the bind address :)

File: hack/local-development/common/helpers
Comment (breuerfelix) [reply hack/local-development/common/helpers]: cc @nschad

Comment (nschad) [reply hack/local-development/common/helpers]: I forgot to remove it, the line below the TODO can be optimized by grouping instead of many subshells.
Docs: https://www.shellcheck.net/wiki/SC2235

Not really necessary you can remove the comment if you want.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (rfranzke) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: Thanks, please add a comment with a link to this issue so that we don't forget about that.

File: hack/local-development/common/helpers
Comment (rfranzke) [reply hack/local-development/common/helpers]: 😵‍💫 Either fix it now or remove it, I doubt that anybody will pick this TODO up otherwise 😉

File: charts/images.yaml
Comment (rfranzke) [inline charts/images.yaml]: As discussed in the other PR, we probably should not rely on the GCR mirror and rather push these images to the `gardener` registry.

File: pkg/features/features.go
Comment (rfranzke) [reply pkg/features/features.go]: Would it be an option to 
- move this function to https://github.com/gardener/gardener/tree/master/pkg/utils/validation/cidr (possibly the `cidr` package could be renamed to `networking` or similar)
- pass a `bool` to the function describing whether IPv6 single stack should be allowed?

Then the respective components could pass their feature gates accordingly to the `bool` argument?

**Comment by timebertt** [comment]
/assign
I will take over the PR from @breuerfelix

File: charts/images.yaml
Comment (timebertt) [reply charts/images.yaml]: tracked in https://github.com/gardener/ci-infra/issues/619

Comment (timebertt) [reply charts/images.yaml]: The image switch can be done in a dedicated PR to unblock this one.
Continuing with docker hub images is fine – at least for now. The local kind seed basically has dual-stack internet access and is able to pull images from docker hub.
Later on, when working on the IPv6 shoot workers this won't work, but we will finish https://github.com/gardener/ci-infra/issues/619 by then probably.

Hence, I will drop the related changes from this PR.

File: pkg/features/features.go
Comment (timebertt) [reply pkg/features/features.go]: There were some misunderstandings here: with the change we use a central feature gate map shared across all component, but the components still register their individual feature gates.
This doesn't change that feature gates are only available on the components for which they are relevant.
However, we can now share code that checks a certain feature gate's state across components, which wasn't possible before.

I explained this in a code comment in a new commit :)

**Comment by timebertt** [comment]
I resolved the remaining comments and all merge conflicts.
Though, I still need to iterate over this once more, clean some things up, and do a final test.
Until that is done:
/hold

**Comment by timebertt** [comment]
/hold cancel
@rfranzke @ScheererJ thanks for your patience and all your feedback so far.
I've rebased the PR, addressed the feedback, introduced a workaround for https://github.com/kubernetes-sigs/kind/issues/3114, and simplified some things again (see the last 11 commits).

This is ready for another round of review now :)

**Comment by timebertt** [comment]
I did some final tests and had to fix some more stuff in provider-local.
With the latest changes, bringing up an IPv6 single-stack seed works with both the skaffold-based and process-based setup. :)

**Comment by timebertt** [comment]
Rebase done :)

**Comment by timebertt** [comment]
Rebased again :)

**Comment by timebertt** [comment]
@rfranzke @ScheererJ Could you kindly take another look before the rebase fun starts again? 😉

File: example/20-componentconfig-gardenlet.yaml
Comment (ScheererJ) [inline example/20-componentconfig-gardenlet.yaml:117]: I am a bit confused why @rfranzke's comment was resolved while the feature gate is still enabled here. It looks like the result of the discussion was the opposite. Can you please clarify?

File: example/gardener-local/gardenlet/values.yaml
Comment (ScheererJ) [inline example/gardener-local/gardenlet/values.yaml:45]: I am a bit confused why @rfranzke's comment was resolved while the feature gate is still enabled here. It looks like the result of the discussion was the opposite. Can you please clarify?

File: example/gardener-local/controlplane/auth-webhook-kubeconfig-local-ipv6.yaml
Comment (ScheererJ) [inline example/gardener-local/controlplane/auth-webhook-kubeconfig-local-ipv6.yaml:8]: Is this supposed to be the real address? Why not use the IPv6 documentation range instead?

File: example/gardener-local/controlplane/auth-webhook-kubeconfig-skaffold-ipv6.yaml
Comment (ScheererJ) [inline example/gardener-local/controlplane/auth-webhook-kubeconfig-skaffold-ipv6.yaml:8]: Is this supposed to be the real address? Why not use the IPv6 documentation range instead?

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (ScheererJ) [inline example/gardener-local/gardenlet/values-ipv6.yaml]: Any particular reason for using fairly readable/easy ranges for everything except for the `nodes`? Would it not make sense to have the `nodes` range readable/easy as well?

File: example/provider-local/seed-kind/local-ipv6/patch-seed.yaml
Comment (ScheererJ) [inline example/provider-local/seed-kind/local-ipv6/patch-seed.yaml]: Again, it is unclear why we end up with hand-crafted, readable ranges for everything except to `nodes`. Could you please clarify?

File: hack/kind-up.sh
Comment (ScheererJ) [inline hack/kind-up.sh:191]: Would it not be better to simply adapt https://github.com/gardener/gardener/blob/master/charts/gardener/provider-local/extension/templates/coredns/custom-configmap.yaml? Just adding another override in the custom dns configuration feels like the straight forward way to do it. It would also remove the need to trigger the rollout.

File: hack/ci-common.sh
Comment (ScheererJ) [inline hack/ci-common.sh:42]: Could you please share how this is related to this PR? I do not see the connection.

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Thanks for the changes so far. I have a few questions/comments.

File: example/20-componentconfig-gardenlet.yaml
Comment (timebertt) [reply example/20-componentconfig-gardenlet.yaml:117]: That's my fault. I missed adding the results of my out-of-band discussion with @rfranzke on github.

The reasoning is that the `IPv6SingleStack` feature gate doesn't change apiserver's/gardenlet's default behavior. It only allows the creation of IPv6 seeds and shoots. The behavior only changes when creating such objects.
We agreed on enabling it by default in the local setup so that devs don't have to modify the configuration to get started with IPv6. If devs don't want to use IPv6, nothing changes.
Hence, I clarified the implications of enabling the feature gate in `feature_gate.md`.

File: example/gardener-local/controlplane/auth-webhook-kubeconfig-local-ipv6.yaml
Comment (timebertt) [reply example/gardener-local/controlplane/auth-webhook-kubeconfig-local-ipv6.yaml:8]: Yes, this is the IP of the `gardener-admission-controller` service in the namespace.
I added the same comment as in the other file to make this easier to understand.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (timebertt) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: The `nodes` CIDR is determined by the range of the `kind` docker network. This is configured by kind and we don't want to make it more complicated by requiring devs to recreate the `kind` network with a "readable" range.

File: hack/kind-up.sh
Comment (timebertt) [reply hack/kind-up.sh:191]: Unfortunately, the custom coredns is not configured as the default DNS server and is hence only used by some of the pods, see the `dnsconfig` webhook of provider-local: https://github.com/gardener/gardener/blob/fd601cb65732d7c5efa35e4736e064a1f21ce4f5/pkg/provider-local/webhook/dnsconfig/add.go#L72-L78

The motivation for the new `garden.local.gardener.cloud` hostname was to make it resolvable/reachable from everywhere. Using the custom coredns doesn't fulfill that requirement, unfortunately.

File: hack/ci-common.sh
Comment (timebertt) [reply hack/ci-common.sh:42]: It's not strictly related to this PR.
However, I debugged an e2e test failure and was wondering whether the new `containerd-configuration-local-setup` was properly executed on the shoot worker node. Then I noticed, that we don't dump the logs yet, so I added it :)

**Comment by timebertt** [comment]
@ScheererJ I replied to your questions, I hope things are clearer now. Can you take another look please? :)

**Comment by timebertt** [comment]
.... aaaaaaaand rebased 😁

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (ScheererJ) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: Would it not be possible to automate the creation of the `kind` network and still achieve nice IP addresses?
I completely agree that additional manual steps should be omitted wherever possible, but the `kind` cluster is created by "us". Therefore, we should be able to configure it properly.
WDYT?

File: hack/kind-up.sh
Comment (ScheererJ) [reply hack/kind-up.sh:191]: Everywhere meaning everywhere within the cluster. Right?
If I resolve `garden.local.gardener.cloud` on my machine I will not get a proper response.

Nevertheless, you are correct in that the custom coredns config does not solve the issue. We just have too many dns servers in the local setup ;-) ...

File: cmd/gardenlet/app/app.go
Comment (rfranzke) [inline cmd/gardenlet/app/app.go:137]: Nit: Doesn't this `SetFromMap` invocation still better belong here than in `validate`?

**Review by rfranzke — COMMENTED** [review]
Looks good

File: hack/kind-up.sh
Comment (timebertt) [reply hack/kind-up.sh:191]: Correct. Currently, it does not resolve on the host machine, because it was not needed so far.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (timebertt) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: Fair enough. We also tell developers to recreate the kind network if it gets created with the "wrong" IPv4 CIDR (ref https://github.com/gardener/gardener/pull/6935). This could be automated as well.
I will give it a try and add a new commit to this PR to tackle both problems.

File: cmd/gardenlet/app/app.go
Comment (timebertt) [reply cmd/gardenlet/app/app.go:137]: Calling `SetFromMap` in `run` is too late. The config file validation will complain about the IPv6 `seedConfig` if the `IPv6SingleStack` feature gate is not enabled (via `SetFromMap`) beforehand.
I improved the comment to make this clearer and performed the change consistently across all component entry points.

File: example/gardener-local/gardenlet/values-ipv6.yaml
Comment (timebertt) [reply example/gardener-local/gardenlet/values-ipv6.yaml]: Done in a23065d9514f69f2b42f3bc0ef859fd05ffa05d8, let me now what you think :)

**Comment by timebertt** [comment]
@rfranzke @ScheererJ I rebased and addressed your suggestions :)

**Comment by timebertt** [comment]
This one keeps being a pain to rebase. @rfranzke @ScheererJ I would highly appreciate a relieving approval 😄

File: cmd/gardener-controller-manager/app/options.go
Comment (rfranzke) [inline cmd/gardener-controller-manager/app/options.go]: Shouldn't we rather move this to the `app.go` before the `opts.complete()` is called? It seems weird/wrong to have it here in `validate()`, don't you agree?

Comment (timebertt) [reply cmd/gardener-controller-manager/app/options.go]: > Calling `SetFromMap` in `run` is too late. The config file validation will complain about the IPv6 `seedConfig` if the `IPv6SingleStack` feature gate is not enabled (via `SetFromMap`) beforehand.

ref https://github.com/gardener/gardener/pull/7561#discussion_r1155738146

Comment (rfranzke) [reply cmd/gardener-controller-manager/app/options.go]: I am not suggesting to call it in `run()` but before `opts.complete()` which is called here: https://github.com/gardener/gardener/blob/7b30c1147001d7ceaab6b822ad171be26e4cd12f/cmd/gardener-controller-manager/app/app.go#L62

Comment (timebertt) [reply cmd/gardener-controller-manager/app/options.go]: Ah sorry, I misread your comment.
You're right., `SetFromMap` should rather be done in `complete()` than in `validate()`.
Calling it before `complete()` is too early, the config is only decoded in `complete()`.

I will move the `SetFromMap` calls to `complete()`.

Comment (timebertt) [reply cmd/gardener-controller-manager/app/options.go]: Done in a new commit, tell me if you like it :)

**Review by ScheererJ — COMMENTED** [review]
/lgtm

**Comment by nschad** [comment]
🥳 🕺 💃 

Good job everyone!

**Comment by jguipi** [comment]
It seems like since that PR was merged, there are a few errors that are generated by the remote-local-setup : 

while trying to start the Gardelet ( `make start-gardenlet SEED_NAME=local` ) 
image

Putting the emphasis on : 
```
panic: failed during bootstrapping: incorrect node network specified in seed configuration (cluster node="172.18.0.2" vs. config="fd00:10::/64")
```

And while trying to run ( ` KUBECONFIG=${KUBECONFIG%:*} USER=root make start-extension-provider-local` ). I'm getting the following.

image

Can you look at that potential regression ?

**Pull request merged**
